Transitive packages and vulnerability indicators in Solution PM UI

martinrrm opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->

# Bug

<!-- If this is an engineering change or test change only, you do not need an issue. -->
<!-- Find or create an issue in NuGet/Home and paste the full url. -->
<!-- At the maintainers discretion, multiple changes may apply to a single issue, but only when the PRs are all created within a short period of time. -->
Fixes: https://github.com/NuGet/Home/issues/13216

## Description
This feature brings to the Solution-level PM UI the Transitive Packages experience that we previously added to the Project-level PM UI. Additionally, we added new UX to the projects / versions grid in the Details Pane to show the package level of the selected package, as well as any vulnerabilities.

Accessibility localization PR https://github.com/NuGet/NuGet.Client/pull/6051

![image](https://github.com/user-attachments/assets/c770e616-f044-4207-a995-62fb41134759)

### Accessibility
![image](https://github.com/user-attachments/assets/5f389915-e5e3-4640-babc-8ede07088fba)

#### Details pane
It has a minimum size so the columns will always look "good"

![image](https://github.com/user-attachments/assets/b64d2865-c158-48dc-85e0-c45b79173fb7)

## PR Checklist

- [x] Meaningful title, helpful description and a linked NuGet/Home issue
- [x] Added tests
- [x] Link to an issue or pull request to update docs if this PR changes settings, environment variables, new feature, etc.

jebriede committed:
First step to fetch transitive packages in Solution PM UI package list and installed tab vulnerability warning (#5994)

First step to show transitive packages in Solution PM UI and show vulnerability warnings in Solution PM UI Installed Tab

martinrrm committed:
Get the packages metadata for all installed packages in the solution level (#5998)

* Get all the installed packages metadata in the Solution level

jebriede committed:
Represent all transitive origins for all installed package versions in tooltip (#6021)

martinrrm committed:
[Transitive's in Solution] Vulnerability indicators in solution level details pane for top-level packages (#6010)

* Vulnerability indicators for top level in solution

martinrrm committed:
[Transitive's in Solution] Transitive's data in Solution Details Pane (#6034)

* Show transitive's data in Details Pane

jebriede committed:
Show transitive package details for packages that are top-level and transitive in the solution (#6040)

martinrrm committed:
[Transitive's in Solution PM UI] Update vulnerability warning tooltips (#6043)

* Update vulnerability warning tooltips

kartheekp-ms commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/IsGreaterThanToVisibilityConverter.cs:
```diff
@@ -9,16 +9,20 @@
 namespace NuGet.PackageManagement.UI
 {
-    public class DownloadCountToVisibilityConverter : IValueConverter
+    public class IsGreaterThanToVisibilityConverter : IValueConverter
```
For some reason, I am having a hard time understanding the name of the class :) If my understanding is correct, the visibility of some control in the UI is dependent on the count. If the count is greater than the some threshold the control will be enabled. Is that correct? Maybe we can try VS Copilot's smart rename feature.

jebriede replied:
You are correct, @kartheekp-ms. The Converter checks if the value is greater than a threshold to determine visibility. Perhaps GreaterThanThresholdToVisibilityConverter would be a better name? Thoughts?

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml:
```diff
@@ -305,23 +314,115 @@
+                  <TextBlock
+                    VerticalAlignment="Bottom"
+                    Text="{Binding InstalledVersion, Converter={StaticResource VersionToStringConverter}}" />
+                  <imaging:CrispImage
```
Could the warning icon be put before the installed version? This would reduce visual clutter by aligning the icons. Also, it's more consistent with the positioning in the packages list (warning icon on the left of installed version).


_rough mockup (left is from this PR, right is my suggestion):_
![image](https://github.com/user-attachments/assets/4ee5edd1-3601-4d9f-a5a6-47c45da17236)

martinrrm replied:
When I presented this to the UX board the warning icons where on the left* side, but their suggestion was to move to the right* side so it matches the Installed tab warning icon at the top

jebriede replied:
@martinrrm do you mean vice versa? The icons are currently on the right side and I believe when you presented this initially you had them on the left and the advice was to put the warning icons on the right side. Please correct me if I'm wrong.

martinrrm replied:
Mb, I corrected the message, but yeah they suggested to change it so both installed "headers" have the icon at the same place, and have them mixed (rows with icon to the left and header to the right) would look weird

donnie-msft replied:
Ok, in that case, the Installed tab could also have probably been changed. We still have 2 different approaches, again, the packages list shows on the left.

![image](https://github.com/user-attachments/assets/e066c196-c7a1-4fa8-a7ea-ad28dfab8bdc)

martinrrm replied:
Thanks Donnie, that a great suggestion, we have been getting UX approval for specific features and not considering the PM UI consistency. We should probably get back to the UX board and look at the whole PM UI. During my last UX meeting we discussed a lot of opportunities we could do improve the PM UI experience, I'll send a link to the meeting recording.

donnie-msft replied:
Look forward to seeing that! In the meantime, if this is "ok" with UX board, then I think it could be easily changed later and not be too jarring to customers.

jgonz120 replied:
Not for this PR, but could it make sense to move the vulnerability information to a column of its own? This would help align the icons and we could also add additional icons to denote the max severity of the vulnerabilities. 

✅ - No Vul
⚠️ - Low/Moderate/High
🔴 - Critical 
❔ - Unkown

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/SolutionView.xaml.cs:
```diff
@@ -239,10 +240,21 @@ private void ListView_SizeChanged(object sender, SizeChangedEventArgs e)
+            var width = _projectList.ActualWidth - 8;
+
+            // If there are more than 7 projects the scrollbar will be visible
+            if (_projectList.Items.Count > 7)
```
Could the screenshots from testing be added to this PR so we can see how this works with different resolutions/etc? I'm not sure if we need PLOC testing, but maybe ensuring the minimum resolution supported looks as expected would be good enough.

martinrrm replied:
![image](https://github.com/user-attachments/assets/3f12b515-487b-4841-aa92-9be25d8d10d3)

martinrrm replied:
Above image is how it looks with a low resolution. Details pane has a minimum width so windows size doesn't really affect the table content.

donnie-msft replied:
Great! I think this should be in the PR description as well.

martinrrm replied:
https://github.com/NuGet/NuGet.Client/pull/6051
Realized that there was missing localization and narrator

martinrrm replied:
Done! Added this and accessibility testing to the PR description

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -29,16 +29,84 @@ public override async Task<SearchResult<IPackageSearchMetadata>> ContinueSearchA
+
+        private static async Task<IPackageSearchMetadata[]> GetRemainingPackagesMetadataAsync(IEnumerable<PackageCollectionItem> packages, FeedSearchContinuationToken searchToken, CancellationToken cancellationToken)
+        {
+            var remainingPackages = packages
```
I believe the installed packages are already grouped by ID here: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Common/PackageCollection.cs#L64

jebriede replied:
The grouping in PackageCollection is grouping by PackageIdentity which is an object that consists of the Id (name of package) and Version. Here, we are grouping by the more specific Id only to sort the versions per package Id and skip the first one.

donnie-msft replied:
> by the more specific Id 

I'm not following what's more specific? The package ID should be unique.

jebriede replied:
Grouping by PackageIdentity will group on PackageId + PackageVersion.

Here we are grouping on PackageId only so we can get a list of PackageVersions grouped by PackageId.

As an example if we have Newtonsoft.Json versions 13.0.0 and 12.0.0 installed across the solution, there will be two PackageIdentity objects like this:
PackageIdentity { Id = "Newtonsoft.Json", Version = "13.0.0" }
PackageIdentity { Id = "Newtonsoft.Json", Version = "12.0.0" }

When we group by Id here, we will group all Newtonsoft.Json Ids and have a list of the Versions [ "13.0.0", "12.0.0" ]. Then, we sort them and skip the first one per each package Id and process the rest.

donnie-msft commented on test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Feeds/InstalledAndTransitivePackageFeedTests.cs:
```diff
@@ -232,8 +232,13 @@ public async Task SearchAsync_WhenMultitargetingAndEmptySearch_ReturnsOnePackage
                 // Returns latest transitive package version
-                e => Assert.Equal(new PackageIdentity("transitivePackageC", new NuGetVersion("0.0.2")), e.Identity));
+                e => Assert.Equal(new PackageIdentity("transitivePackageC", new NuGetVersion("0.0.2")), e.Identity),
+                // Returns all the transtive packages that are not the latest verion
```
Could the comment be expanded to explain why this assertion is made?

```suggestion
                // Returns all the transitive packages that are not the latest version because.....
```

martinrrm replied:
So this PackageCollection is sorted, first we display the Installed Packages and then the TransitivePackages, and in each "group" they are sorted by latest. 

In this test I'm trying to test that they are sorted as expected

donnie-msft replied:
If they are sorted, why is transitivePackageC 0.0.2 asserted before version 0.0.1 ?

Are you planning to fix the typos I suggested?

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs:
```diff
@@ -136,6 +137,28 @@ public string ByOwnerOrAuthor
+            set
+            {
+                _vulnerableVersions = value;
+                OnPropertyChanged(nameof(VulnerableVersions));
```
Is this necessary? I don't see where the entire dictionary is being replaced, so I don't think it's being used.

jebriede replied:
Good point. I think we can delete the Setter altogether since the Dictionary is instantiated and only added to and this setter is never used. Then, the private _vulnerableVersions can be readonly.

jebriede replied:
@donnie-msft I addressed this now in the latest commit. I moved the OnPropertyChanged to a place where we add items to the VulnerableVersions dictionary, and also added an OnPropertyChanged for the VulnerableVersionsString which was previously missing. I removed the setter here because it was not being used. Thank you!

donnie-msft reviewed (commented):
Great to see this being added, nice work!
I'd like to see some accessibility testing to ensure we're not introducing a11y bugs for the new UI elements (https://github.com/NuGet/NuGet.Client/blob/dev/docs/ui-guidelines.md#accessibility-testing).

My main UI behavior concern is whether the calculations for columns and scrolling works well at different resolutions.

More unit tests to cover more of the package feeds and Services logic changes would be good as well.

jebriede commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageItemLoader.cs:
```diff
@@ -262,85 +263,105 @@ public IEnumerable<PackageItemViewModel> GetCurrent()
-                    {
-                        versionOverride = versionOverrides[0];
-                    }
+                    existingListItem.UpdateInstalledPackagesVulnerabilities(new PackageIdentity(packageId, packageVersion));
```
@martinrrm Am I remembering correctly that we changed this to accept metadataContextInfo.Identity rather than instantiate a new PackageIdentity? I'm wondering how this made its way back here.

jebriede replied:
I fixed this in the latest commit.

jgonz120 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/PackageSolutionDetailControlModel.cs:
```diff
@@ -181,18 +211,20 @@ private async Task UpdateInstalledVersionsAsync(CancellationToken cancellationTo
         /// It is likely not called more than once in an action.
         /// </summary>
-        private async Task<IPackageReferenceContextInfo> GetInstalledPackageAsync(
+        private async Task<(IPackageReferenceContextInfo TopLevelPackage, IPackageReferenceContextInfo TransitivePackage)> GetInstalledAndTransitivePackagesAsync(
```
Trying to make sure I'm understanding things here, this is returning the package information for a project and a specific packageId?

Is it possible for a single packageId to be both a toplevel and transitive package for a project? If not then I think this should be renamed to `GetInstalledOrTransitivePackageAsync`. 

`private async Task<(IPackageReferenceContextInfo Package, bool IsTransitive)> GetPackageContextAsync`

jebriede replied:
@jgonz120 It's not possible for a package to be both top-level and transitive in a single project. It's only possible for a package to be top-level and transitive in a across the solution, i.e. top-level in one project, transitive in another.

@martinrrm I think the suggested rename is a good idea because it looks like we're getting package information for a project/package combo. Could you confirm this is what we're doing and what you think of the proposed rename?

jgonz120 replied:
I put this `private async Task<(IPackageReferenceContextInfo Package, bool IsTransitive)> GetPackageContextAsync` and didn't explain it.

If only one package will ever be returned we can simplify the return with a bool to indicate if the package found is transitive or top level.

jgonz120 replied:
Or even `private async Task<(IPackageReferenceContextInfo Package, PackageLevel PackageLevel)> GetPackageContextAsync`

jebriede replied:
@jgonz120: I created https://github.com/NuGet/NuGet.Client/pull/6065 to incorporate your feedback. Please let me know if that's what you had in mind and leave any feedback on that PR. Thanks!

jgonz120 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageInstallationInfo.cs:
```diff
@@ -49,6 +50,26 @@ private async ValueTask InitializeAsync(CancellationToken cancellationToken)
         private string _versionRequested;
 
+        private int _installedVersionMaxVulnerability = -1;
+        public int InstalledVersionMaxVulnerability
```
I think I might have already asked this, what does this mean?😅

jebriede replied:
@jgonz120 it corresponds with:
https://github.com/NuGet/NuGet.Client/blob/1e0115a608073be2ad615201be30b93dd628300d/src/NuGet.Core/NuGet.Protocol/Model/PackageVulnerabilitySeverity.cs#L6

jgonz120 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs:
```diff
@@ -854,22 +956,22 @@ public void UpdatePackageStatus(IEnumerable<PackageCollectionItem> installedPack
             OnPropertyChanged(nameof(Status));
         }
 
         public void UpdateTransitivePackageStatus(NuGetVersion installedVersion)
```
Would it make sense for the PackageItemViewModel to keep track of whether or not the package is transitive?

jebriede replied:
It does in PackageLevel. I know, this method should not be public and the caller should not need to call the appropriate UpdatePackageStatus/UpdateTransitivePackageStatus method, and we should instead call UpdatePackageStatus and the ViewModel should do the right thing depending on whether it's top-level or transitive. However, this code existed already and at this point I want to make as few changes as possible so we can complete this work and make that enhancement later.

jgonz120 replied:
Sounds good to me!

jgonz120 commented on src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs:
```diff
@@ -839,6 +911,36 @@ private async Task ReloadPackageMetadataAsync()
+            IEnumerable<PackageVulnerabilityMetadataContextInfo> vulnerabilityInfoList =
+                        await _vulnerabilityService.GetVulnerabilityInfoAsync(packageIdentity, cancellationToken);
+
+            SetVulnerabilityMaxSeverity(packageIdentity.Version, vulnerabilityInfoList?.FirstOrDefault()?.Severity ?? -1);
```
If you're getting a list of vulnerabilities, would it make sense to select the one with the highest severity, instead of just the first, for setting the max severity?

martinrrm replied:
So, I wasn't able to locate where we explain this, but my understanding is that they are already sorted by severity when we query them from the database so there is no need to "sort" again.

[This code was just moved into another function](https://github.com/NuGet/NuGet.Client/pull/6044/files/85f658abf961201b1ba8dea5509aa913e61b2774#diff-118d6d1c3dcb2549a9436b8dafadcf915e0ae147a2ed54f3d82767252ee94701L817), but the implementation is the same and is being used in project PM UI too.

jgonz120 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs:
```diff
@@ -29,16 +29,84 @@ public override async Task<SearchResult<IPackageSearchMetadata>> ContinueSearchA
+            IEnumerable<PackageCollectionItem> transitivePkgsWithOrigins = GetTransitivePackagesWithOrigins(_transitivePackages);
+
+            // Get the metadata for the latest version of the installed packages and transitive packages
+            IEnumerable<PackageCollectionItem> installedPackagesLatest = GetLatestPackages(_installedPackages);
```
Nit: can this be updated to match the other variable name? `latestInstalledPackages`

jebriede committed:
Code cleanup and addressing PR feedback.

jgonz120 reviewed (dismissed):
Looks good to me!

martinrrm committed:
Update ProjectInstallationInfo ToString() on Details Pane table and improve localization of table headers (#6051)

kartheekp-ms commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/GreaterThanThresholdToVisibilityConverter.cs:
```diff
@@ -9,16 +9,20 @@
+                if (value is not null)
                 {
-                    return Visibility.Visible;
+                    long intValue = System.Convert.ToInt64(value, culture);
```
Convert.ToInt64() will throw an exception if the value is not a valid long value. The example in the [doc](https://learn.microsoft.com/dotnet/api/system.convert.toint64?view=net-8.0#system-convert-toint64(system-object)) has catch blocks for various exceptions this method can throw.

jgonz120 replied:
I'm guessing this is done so an object or a string can be converted to the correct value. Would it worth creating a `bool TryConvertToInt64(obj, formatter, out long? convertedValue)` method somewhere?

jebriede replied:
@kartheekp-ms thanks for the feedback. @jgonz120 I like this suggestion. I've got this ready and will push changes in a moment.

jebriede replied:
@kartheekp-ms @jgonz120 pushed an update that addresses this case and has a new test for invalid conversion scenarios.

kartheekp-ms commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/GreaterThanThresholdToVisibilityConverter.cs:
```diff
@@ -9,16 +9,20 @@
+                if (value is not null)
                 {
-                    return Visibility.Visible;
+                    long intValue = System.Convert.ToInt64(value, culture);
```
nit: The variable name `intValue` is not clear to me. Are we referring to the initial value?

jebriede replied:
@kartheekp-ms do you have a suggestion for another name?

jebriede replied:
I am making it `initialValue` and `thresholdValue`

jebriede committed:
- Adding exception handling for converter and test for invalid conversion cases.
- Only adding to VulnerableVersions dictionary if it does not already exist because the same version is installed in another project and only if it's added should we fire OnPropertyChanged.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs:
```diff
@@ -136,6 +138,17 @@ public string ByOwnerOrAuthor
+            }
+        }
+
+        private readonly Dictionary<NuGetVersion, int> _vulnerableVersions = [];
+        public Dictionary<NuGetVersion, int> VulnerableVersions => _vulnerableVersions;
```
Is anything actually listening to VulnerableVersions ? Maybe I'm missing it in a binding somewhere?

jebriede replied:
It's used here: https://github.com/NuGet/NuGet.Client/pull/6044/files#diff-5c15ee840d9063ec5f345761223159c7ecde0f058e0e0190e359adcfee761290R170

It was also supposed to be used in the XAML for the VulnerableVersions.Count but there's typo. I'm fixing the XAML binding now.

jebriede replied:
Fixing in: https://github.com/NuGet/NuGet.Client/pull/6056

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs:
```diff
@@ -136,6 +138,17 @@ public string ByOwnerOrAuthor
+            }
+        }
+
+        private readonly Dictionary<NuGetVersion, int> _vulnerableVersions = [];
```
Is this field being used? I think it doesn't need both the field & property.

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs:
```diff
@@ -839,6 +901,41 @@ private async Task ReloadPackageMetadataAsync()
+            {
+                if (VulnerableVersions.TryAdd(version, maxSeverity))
+                {
+                    OnPropertyChanged(nameof(VulnerableVersions));
+                    OnPropertyChanged(nameof(VulnerableVersionsString));
```
nit: Ideally, a CollectionChanged event could be used to fire the appropriate OnPropertyChanged events when the collection gets changed.

jebriede replied:
That's a really good point. This should be an observable set / dictionary that fires CollectionChanged when it gets updated, and we listen to that and fire any additional OnPropertyChanged as needed.

kartheekp-ms commented:
Looking at the screenshot mentioned in the PR description, I think `Uninstall` button should be invisible for transitive packages. Only `Install` button might be applicable because the install action promotes a transitive dependency to top-level dependency. I am unsure if this feedback is out of scope for this PR.

![image](https://github.com/user-attachments/assets/c770e616-f044-4207-a995-62fb41134759)

donnie-msft commented on src/NuGet.Clients/NuGet.PackageManagement.UI/PackageItemLoader.cs:
```diff
@@ -262,85 +262,105 @@ public IEnumerable<PackageItemViewModel> GetCurrent()
             }
 
-            var listItemViewModels = new List<PackageItemViewModel>(capacity: _state.ItemsCount);
+            var listItemViewModels = new Dictionary<string, PackageItemViewModel>();
```
I'm trying to understand what the dictionary is being used for. Could someone explain how the same package would exist in `_state.Results.PackageSearchItems` multiple times?  Ideally, a comment would be good here as well.

jebriede replied:
@donnie-msft the changes made here: https://github.com/NuGet/NuGet.Client/pull/6044/files#diff-b1b2dd9d0be30f84d9309d08db86156cff27b5df8e43b54c13e90bfa5fbf72f9R38-R43 are related to this.

Let's see if I can explain this in a way that makes sense in a comment. :-) (This is actually the main reason I did the live code review meeting to explain this part). The Solution PM UI needs to know about all the versions installed across the solution, in each project, not just the latest version. For the Project PM UI, this will always only be the latest version because of how restore works. But limiting it to just the latest doesn't work in the solution because we can have the same package installed at different versions in different projects and since one of those can be vulnerable, not just the latest, we need to know all of the package versions for each package. As such, the `InstalledAndTransitivePackageFeed` will now return all the package versions. However, we want to create a single PackageItemViewModel so it shows up only once in the packages list. We keep track of the PackageItemViewModels we've created for that Package Id (the string key) and retrieve the view model if it was already created, so we can add the additional versions to it.

I hope this explanation helps! I can probably explain it better in real time so if you have questions or want a deeper dive, please set up a quick meeting and we'll go over this in more detail!

donnie-msft reviewed (approved):
Great work and can't wait to see how customers use this! 
Left some questions and nits.

jebriede commented:
> Looking at the screenshot mentioned in the PR description, I think `Uninstall` button should be invisible for transitive packages. Only `Install` button might be applicable because the install action promotes a transitive dependency to top-level dependency. I am unsure if this feedback is out of scope for this PR.

@kartheekp-ms We already fixed this. I thought the PR was completed but it looks like it's not merged yet. This fixes the issue: https://github.com/NuGet/NuGet.Client/pull/6050. I'll ask @martinrrm to retarget dev and submit this fix after this PR goes in.
